Testing BA-MBSE Persistence Refactor Merge

HuiJun opened the pull request:
deconflicted branch of https://github.com/Open-MBEE/exec-mms/pull/235

BA-MBSE committed:
Decouples Federated Persistence layer from controller and service layers.
Introduces a unified persistence layer interface with a federated implementation.
Added msosa, view, and oauth modules

BA-MBSE committed:
Remove redundant commit save

BA-MBSE committed:
CodeQL updates

BA-MBSE committed:
More CodeQL changes

BA-MBSE committed:
Added timeout to database delete statement

BA-MBSE committed:
Removing overarching transactional annotations

BA-MBSE committed:
Removed transactional annotations from DefaultPermissionService

BA-MBSE committed:
Removed @Transactional annotation from modules that are not tasked with persistence implementation

BA-MBSE committed:
Fixes for failing crud tests

BA-MBSE committed:
Increased request delay for crud tests

BA-MBSE committed:
Working on fixing permissions test failures

BA-MBSE committed:
increased delay in permissions test

BA-MBSE committed:
Fixing problem with bean creation for permissions delegates that was preventing them from getting transaction proxies created

BA-MBSE committed:
Fixed artifacts behavior

BA-MBSE committed:
extending artifacts test delay to 1000 ms

BA-MBSE committed:
Fixing elastic test failures

BA-MBSE committed:
Fixing groups test failures

BA-MBSE committed:
Merge branch 'develop' into ba_persistence_refactor

HuiJun committed:
Merge branch 'ba_persistence_refactor' of https://github.com/BA-MBSE/mms into develop

# Conflicts:
#	crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java

dlamoris commented on cameo/src/main/java/org/openmbee/mms/cameo/services/CameoViewService.java:
```diff
@@ -75,42 +82,47 @@ public void addChildViews(ElementsResponse res, Map<String, String> params) {
+        List<ElementJson> groups = getNodePersistence().findAllByNodeType(projectId, refId, commitId,
+            CameoNodeType.GROUP.getValue());
+
+        ElementsResponse res = this.read(projectId, refId, buildRequestFromJsons(groups), params);
```
groups already is the right list of json

dlamoris replied:
fixed in #260

dlamoris commented on cameo/src/main/java/org/openmbee/mms/cameo/services/CameoViewService.java:
```diff
@@ -75,42 +82,47 @@ public void addChildViews(ElementsResponse res, Map<String, String> params) {
     }
 
     public ElementsResponse getGroups(String projectId, String refId, Map<String, String> params) {
```
add override

dlamoris reviewed (changes requested):
have not went over permissions related changes yet
things to check:
- some performance optimizations in federated classes where it's getting entire commit json instead of just commit info from db
- updating deleted elements, deleted key in json
- series of inits across sub/superclasses during element get/post, check logic if element post has deletes as well as updates

HuiJun commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultCommitService.java:
```diff
@@ -93,56 +69,50 @@ public CommitsResponse getRefCommits(String projectId, String refId, Map<String,
+            commitIds.add(commit.getId());
         }
-        res.getCommits().addAll(commitIndex.elementHistory(elementId, commitIds));
+        res.getCommits().addAll(commitPersistence.elementHistory(projectId, elementId, commitIds));
```
*

HuiJun commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultCommitService.java:
```diff
@@ -93,56 +69,50 @@ public CommitsResponse getRefCommits(String projectId, String refId, Map<String,
     public boolean isProjectNew(String projectId) {
-        ContextHolder.setContext(projectId);
-        List<Commit> commits = commitRepository.findAll();
+        List<CommitJson> commits = commitPersistence.findAllByProjectId(projectId);
```
*

dlamoris replied:
fixed in #260

HuiJun commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+    public FederatedNodeChangeDomain(NodeGetDomain nodeGetDomain, CommitDomain commitDomain,
+            FederatedNodeGetDomain getDomain, NodeDAO nodeRepository, NodeIndexDAO nodeIndex, FederatedElementDomain elementDomain,
+            List<NodeUpdateFilter> nodeUpdateFilters) {
+        super(nodeGetDomain, commitDomain, nodeUpdateFilters);
```
*

dlamoris replied:
fixed in #260

Commit:
Merge commit '4e4e8586b744bc371db774f2b7115647003d2965' into feature/persistence_refactor

Commit:
fix NodePersistence.findById interface argument names

Commit:
fix FederatedNodeChangeDomain superclass constructor call

Commit:
change CommitPersistence.elementHistory interface so DefaultCommitService doesn't need to pass in commitIds (this is so it's not getting json just to get ids as getting entire commit json is expensive for federated persistence)

Commit:
remove redundant groups read from CameoViewService

Commit:
remove posted _childViews handling from CameoViewService, this was handling ve structural changes and has been taken out of ve (5+)

dlamoris commented on core/src/main/java/org/openmbee/mms/core/dao/NodePersistence.java:
```diff
@@ -0,0 +1,35 @@
+
+    NodeChangeInfo commitChanges(NodeChangeInfo nodeChangeInfo);
+
+    NodeGetInfo findById(String elementId, String refId, String commitId, String id);
```
fixed in https://github.com/Open-MBEE/exec-mms/pull/260

dlamoris commented on cameo/src/main/java/org/openmbee/mms/cameo/services/CameoNodeService.java:
```diff
@@ -80,38 +82,37 @@ public ElementsResponse read(String projectId, String refId, ElementsRequest req
 
     @Override
-    public void extraProcessPostedElement(ElementJson element, Node node, NodeChangeInfo info) {
-        node.setNodeType(cameoHelper.getNodeType(element).getValue());
```
this was replaced by https://github.com/Open-MBEE/exec-mms/blob/feature/persistence_refactor/federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java#L140

Commit:
continue if no permission to change project org on project update

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/controllers/projects/ProjectsController.java:
```diff
@@ -113,59 +97,72 @@ public ProjectsResponse createOrUpdateProjects(
+                                if (!mss.hasProjectPrivilege(auth, json.getProjectId(), Privileges.PROJECT_DELETE.name(), false) ||
+                                        !mss.hasOrgPrivilege(auth, json.getOrgId(), Privileges.ORG_CREATE_PROJECT.name(), false)) {
+                                    response.addRejection(
+                                        new Rejection(json, 403, "No permission to move project org"));
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultCommitService.java:
```diff
@@ -93,56 +69,50 @@ public CommitsResponse getRefCommits(String projectId, String refId, Map<String,
-            commitIds.add(commit.getCommitId());
+        List<CommitJson> refCommits = commitPersistence.findByProjectAndRefAndTimestampAndLimit(projectId, refId, null, 0);
+        Set<String> commitIds = new LinkedHashSet<>();
+        for (CommitJson commit: refCommits) {
```
fixed in #260

Commit:
reduce amount of commits gotten to test for new project in DefaultCommitService and CameoCommitService

Commit:
add commitId to node get responses in DefaultNodeService and CameoNodeService,

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -99,187 +57,158 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
         ElementsResponse response = new ElementsResponse();
         response.getElements().addAll(info.getActiveElementMap().values());
         response.setRejected(new ArrayList<>(info.getRejected().values()));
-        response.setCommitId(info.getCommitId());
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -99,187 +57,158 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
 
-            ElementsRequest req = buildRequest(id);
-            return read(projectId, refId, req, params);
+            NodeGetInfo getInfo = nodePersistence.findById(projectId, refId, commitId, id);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -300,52 +229,45 @@ protected ElementsRequest buildRequest(Collection<String> ids) {
-            return objectMapper.writeValueAsString(elementJson);
-        } catch (JsonProcessingException e) {
-            logger.error("Error in toJson: ", e);
+    protected Map<String, ElementJson> convertJsonToMap(List<ElementJson> elements) {
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -300,52 +229,45 @@ protected ElementsRequest buildRequest(Collection<String> ids) {
-    protected static <T> Stream<List<T>> batches(List<T> source, int length) {
-        return IntStream.iterate(0, i -> i < source.size(), i -> i + length)
-            .mapToObj(i -> source.subList(i, Math.min(i + length, source.size())));
+    protected List<ElementJson> filter(List<String> ids, List<ElementJson> orig) {
```
fixed in #260

Commit:
removed duplicated convertJsonToMap function in DefaultNodeService, moved filter to JsonDomain

Commit:
prevent adding change info to commit if element isn't actually being updated

dlamoris commented on json/src/main/java/org/openmbee/mms/json/BaseJson.java:
```diff
@@ -157,6 +158,18 @@ public T setCommitId(String commitId) {
+
+    @SuppressWarnings("unchecked")
+    @JsonProperty(IS_DELETED)
+    public T setIsDeleted(String deleted) {
```
deleted key should at least have _ prefix so it doesn't interfere with mdk and other tools thinking it's a valid key

dlamoris replied:
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+            return;
+        }
+
+        super.processElementUpdated(info, element, existing);
```
fixed in #260

Commit:
fix setting initial commit of added element, prefix deleted key with _

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+        ((FederatedNodeChangeInfo) info).getExistingNodeMap().put(element.getId(), node);
+        super.processElementAdded(info, element);
+
+        node.setInitialCommit(element.getDocId());
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+        commitJson.setId(UUID.randomUUID().toString());
+        commitJson.setDocId(commitJson.getId()); 
+
+        NodeChangeInfo info =  super.initInfo(commitJson, overwrite, preserveTimestamps);
```
fixed in #260

Commit:
fix processElementDeleted

Commit:
use commitDAO interface in FederatedNodeChangeDomain instead of commitPersistence for better performance, change convertJsonToMap to accept collection

Commit:
clean up unused things in FederatedNodeChangeDomain

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+@Component
+public class FederatedNodeGetDomain extends NodeGetDomain {
+
+    private CommitPersistence commitPersistence;
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+        // bulk read existing elements in elastic
+        List<ElementJson> existingElements = nodeIndex.findAllById(indexIds);
+        addExistingElements(nodeGetInfo, existingElements); // handeled in addExistingElements
+        return nodeGetInfo;
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+    protected List<String> getRefCommitIds(Instant time) {
+        List<String> commitIds = new ArrayList<>();
+
+        List<CommitJson> refCommits = commitPersistence.findByProjectAndRefAndTimestampAndLimit(getContext().getProjectId(), getContext().getBranchId(), time, 0);
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+        if(info instanceof FederatedNodeGetInfo) {
+            FederatedNodeGetInfo federatedInfo = (FederatedNodeGetInfo)info;
+            Set<String> elementIds = elements.stream().map(ElementJson::getId).collect(Collectors.toSet());
+            List<Node> existingNodes = nodeDAO.findAllByNodeIds(elementIds);
```
fixed in #260

Commit:
uncomment child views handling so tests run for now

Commit:
if searching by commit id doesn't work, try getting by id directly
this fixes jupyter tests because in jupyter, it's doing a read in the same call as an update - the read now always gets the latest commit - which elastic may not have had time to index and results in an error

Commit:
reduce test delay back to 500

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+            // reqElementMap.put(node.getNodeId(), new ElementJson().setId(node.getNodeId()));
+        }
+        if(!transactedElements.isEmpty()){
+            reqElementMap.putAll(convertJsonToMap(transactedElements.parallelStream().collect(Collectors.toList())));
```
fixed in #260

dlamoris commented on cameo/src/main/java/org/openmbee/mms/cameo/services/CameoNodeService.java:
```diff
@@ -46,22 +42,28 @@ public void setMss(MethodSecurityService mss) {
-        String commitId = params.getOrDefault("commitId", null);
-        ContextHolder.setContext(projectId, refId);
-        NodeGetInfo info = nodeGetHelper.processGetJson(req.getElements(), commitId, this);
+        String commitId = params.getOrDefault(CameoConstants.COMMITID, null);
```
fixed in #260

Commit:
support updating existing ref's json properties, will resurrect if previously deleted, will not allow changing parentRef or parentCommit

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java:
```diff
@@ -77,129 +57,107 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
+            }
+            if (refJson.get().isDeleted()) {
+                //un-delete action
+                branch.setDeleted(false);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java:
```diff
@@ -77,129 +57,107 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
-        Instant now = Instant.now();
-        ContextHolder.setContext(projectId);
-
-        Optional<Branch> branchesOption = this.branchRepository.findByBranchId(branch.getId());
```
fixed in #260

Commit:
fix updating org name

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedOrgPersistence.java:
```diff
@@ -0,0 +1,83 @@
+    public OrgJson save(OrgJson orgJson) {
+        Optional<Organization> organizationOptional = orgDAO.findByOrganizationId(orgJson.getId());
+        if (organizationOptional.isPresent()) {
+            throw new ForbiddenException("org " + orgJson.getId() + " already exists");
```
fixed in #260

Commit:
prevent project update from updating schema

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/controllers/projects/ProjectsController.java:
```diff
@@ -113,59 +97,72 @@ public ProjectsResponse createOrUpdateProjects(
+                Optional<ProjectJson> existingOptional = projectPersistence.findById(json.getProjectId());
+                if(existingOptional.isPresent()) {
+                    //Project exists, should merge the json
+                    json.merge(existingOptional.get());
```
fixed in #260

Commit:
fix setting commit id from commitJson id

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedCommitPersistence.java:
```diff
@@ -0,0 +1,183 @@
+        ContextHolder.setContext(commitJson.getProjectId());
+        Commit commit = new Commit();
+        commit.setComment(commitJson.getComment());
+        commit.setCommitId(commitJson.getDocId());
```
fixed in #260

Commit:
sort commits by descending

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedCommitPersistence.java:
```diff
@@ -0,0 +1,183 @@
+            }
+        });
+        List<CommitJson> commits = commitIndexDAO.findAllById(foundCommitIds);
+        commits.sort(Comparator.comparing(BaseJson::getCreated));
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedCommitPersistence.java:
```diff
@@ -0,0 +1,183 @@
+        Branch b = branchOptional.get();
+        List<Commit> commitList = commitDAO.findByRefAndTimestampAndLimit(b, timestamp, limit);
+        commitList.forEach(commit -> commitIds.add(commit.getCommitId()));
+        List<CommitJson> commits = commitIndexDAO.findAllById(commitIds);
```
is ok

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedCommitPersistence.java:
```diff
@@ -0,0 +1,183 @@
+
+    @Override
+    public List<CommitJson> elementHistory(String projectId, String elementId, Set<String> commitDocIds) {
+        ContextHolder.setContext(projectId);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/controllers/branches/BranchesController.java:
```diff
@@ -94,7 +96,14 @@ public RefsResponse createRefs(
+                    if (existingOptional.isPresent()) {
+                        //Branch exists, should merge the json
+                        branch.merge(existingOptional.get());
+                        res = branchService.updateBranch(projectId, branch);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java:
```diff
@@ -77,129 +57,107 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
+                branch.setDeleted(false);
+            }
+        }
+       return branchPersistence.update(branch);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -300,52 +229,45 @@ protected ElementsRequest buildRequest(Collection<String> ids) {
-        if (commitId != null && !commitId.isEmpty()) {
-            cmjs.setId(commitId);
+        if(commitId != null) {
+            cmjs.setCommitId(commitId);
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+    }
+
+    @Override
+    public void addExistingElements(NodeChangeInfo nodeChangeInfo, List<ElementJson> existingElements) {
```
removed in #260

Commit:
check if commitId is latest for gets and pass in null if it is for performance

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedNodePersistence.java:
```diff
@@ -0,0 +1,260 @@
+                } else {
+                    stream.write(separator.getBytes(StandardCharsets.UTF_8));
+                }
+                Collection<ElementJson> result = getNodeGetDomain().processGetJsonFromNodes(ns, commitId)
```
fixed in #260 
should check if caller can just pass in null instead...otherwise there's still redundant checks to see if commit is latest

Commit:
remove any _deleted flag in element to be updated

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedNodePersistence.java:
```diff
@@ -0,0 +1,260 @@
+        ContextHolder.setContext(projectId, refId);
+        List<ElementJson> elements = new ArrayList<>();
+        elements.add(new ElementJson().setId(elementId));
+        return getNodeGetDomain().processGetJson(elements, commitId);
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/domain/DefaultNodeUpdateFilter.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public boolean filterUpdate(NodeChangeInfo info, ElementJson updated, ElementJson existing) {
+        if (!info.getOverwrite()) {
+            if (Constants.TRUE.equals(existing.getIsDeleted()) || isUpdated(updated, existing, info)) {
```
fixed in #260

Commit:
set returned elements refId to match requested refId

Commit:
set refIds for cameo element gets

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultNodeService.java:
```diff
@@ -99,187 +57,158 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
+            List<ElementJson> nodes = nodePersistence.findAll(projectId, refId, commitId);
+            response.getElements().addAll(nodes);
         }
+        response.getElements().forEach(v -> v.setRefId(refId));
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+            } else if (modified.isAfter(time)) { // latest element is after commit
+                Optional<ElementJson> tryExact = nodeIndex.getByCommitId(commitId, nodeId);
+                if (tryExact.isPresent()) {
+                    addActiveElement(info, nodeId, tryExact.get());
```
fixed in #260, refid is set in nodeServices

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+                    Formats.FORMATTER.format(time), refCommitIds);
+                if (e.isPresent()) { // found version of element at commit time
+                    //TODO determine if element was deleted at the time?
+                    addActiveElement(info, nodeId, e.get());
```
fixed in #260, ref id is set in nodeServices

Commit:
allow update even if previous element not found

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+                added = true;
+            } else if (indexElement == null) {
+                logger.warn("node db and index mismatch on element update: nodeId: " + n.getNodeId() + ", docId not found: " + n.getDocId());
+                info.addRejection(element.getId(), new Rejection(element, 500, "Update failed: previous element not found"));
```
fixed in #260

Commit:
handle if element json not found during gets

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+            ElementJson indexElement = info.getExistingElementMap().get(nodeId);
+
+            if (federatedInfo.getExistingNodeMap().get(nodeId).isDeleted()) {
+                rejectDeleted(info, nodeId, indexElement);
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeGetDomain.java:
```diff
@@ -0,0 +1,206 @@
+            if (!existingNodeContainsNodeId(federatedInfo, nodeId)) { // nodeId not found
+                continue;
+            }
+            ElementJson indexElement = info.getExistingElementMap().get(nodeId);
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedElementDomain.java:
```diff
@@ -0,0 +1,64 @@
+        if (null == elementUtils) {
+            return 0;
+        } 
+        return elementUtils.getNodeType(element).ordinal()+1;
```
ok

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedOrgPersistence.java:
```diff
@@ -0,0 +1,83 @@
+        }
+        Organization organization = new Organization();
+        organization.setOrganizationId(orgJson.getId());
+        organization.setOrganizationName(orgJson.getName());
```
nevermind public is set by permissions

Commit:
remove addChildViews delay

dlamoris commented on view/src/main/java/org/openmbee/mms/view/controllers/VeController.java:
```diff
@@ -96,8 +97,17 @@ public ElementsResponse createOrUpdateViews(
+        ViewService viewService = genericServiceFactory.getServiceForSchema( ViewService.class , getProjectType(projectId));
+        ElementsResponse res = viewService.createOrUpdate(projectId, refId, req, params, auth.getName());
+        // ToDo : the following is just a workaround till the indexing part is synchronized W
+        try {
```
fixed in #260 think this was because the added commit check done during read

dlamoris commented on rdb/src/main/java/org/openmbee/mms/rdb/repositories/branch/BranchDAOImpl.java:
```diff
@@ -34,8 +37,10 @@ public Branch save(Branch branch) {
             ContextHolder.setContext(ContextHolder.getContext().getProjectId(), branch.getBranchId());
-            branchesOperations.createBranch();
-            branchesOperations.copyTablesFromParent(branch.getBranchId(), branch.getParentRefId(), null);
+            if(!Constants.MASTER_BRANCH.equals(branch.getBranchId())) {
```
ok

Commit:
throw error if project can't hard delete

dlamoris commented on rdb/src/main/java/org/openmbee/mms/rdb/repositories/ProjectDAOImpl.java:
```diff
@@ -60,14 +61,17 @@ public Project save(Project proj) {
+        try {
+            projectRepository.findByProjectId(projectId).ifPresent(v -> projectRepository.delete(v));
+        } catch (Exception ex) {
+            logger.error("Could not delete project from project Repository");
```
fixed in #260

dlamoris commented on rdb/src/main/java/org/openmbee/mms/rdb/repositories/ProjectDAOImpl.java:
```diff
@@ -60,14 +61,17 @@ public Project save(Project proj) {
-            projectOperations.deleteProjectDatabase(p);
-        } catch(SQLException ex) {
+            projectOperations.deleteProjectDatabase(projectId);
+        } catch (SQLException ex) {
```
fixed in #260

dlamoris commented on rdb/src/main/java/org/openmbee/mms/rdb/config/DatabaseDefinitionService.java:
```diff
@@ -132,12 +141,6 @@ public void generateProjectSchemaFromModels(Project project) throws SQLException
             .createOnly(EnumSet.of(TargetType.DATABASE),
                 metadata.getMetadataBuilder().build());
-
-        try (Connection conn = crudDataSources.getDataSource(project).getConnection()) {
```
ok, master creation moved to default project service

Commit:
update project permissions after changing orgs if needed

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/controllers/projects/ProjectsController.java:
```diff
@@ -113,59 +97,72 @@ public ProjectsResponse createOrUpdateProjects(
+                                        new Rejection(json, 403, "No permission to move project org"));
+                                }
+                                if (projectPersistence.inheritsPermissions(projectJson.getProjectId())) {
+                                    permissionService.setProjectInherit(false, json.getProjectId());
```
fixed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultProjectService.java:
```diff
@@ -77,91 +55,64 @@ public ProjectJson create(ProjectJson project) {
         }
-        throw new InternalErrorException("Could not update project");
+
+        return projectPersistence.update(project);
```
org move permission updates are in controller

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedProjectPersistence.java:
```diff
@@ -0,0 +1,226 @@
+        }
+
+        if(!message.isEmpty()) {
+            throw new NotFoundException(message);
```
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/domain/FederatedNodeChangeDomain.java:
```diff
@@ -0,0 +1,281 @@
+                continue;
+            }
+            Node node = federatedInfo.getExistingNodeMap().get(nodeId);
+            ElementJson indexElement = info.getExistingElementMap().get(nodeId);
```
fixed in #260

Commit:
continue with element delete even if json not found

Commit:
Merge commit 'f7ac4a8af45c3f3347eec2a2b4d0e9a5fe12f144' into fix/persistence_refactor_fixes

merging from develop, deconflict branch from commit

Commit:
add more tests for branching behavior

Commit:
Merge commit '83ec1f6db8a540e1c8bade294c49fa62a62cec25' into fix/persistence_refactor_fixes

dlamoris commented on core/src/main/java/org/openmbee/mms/core/services/DefaultPermissionService.java:
```diff
@@ -357,72 +331,33 @@ private Organization getOrganization(String orgId) {
-                    Branch b = new Branch(getProject(projectId), branchId, false);
-                    branchRepo.save(b);
-                    return b;
+                    RefJson b = new RefJson();
```
before was only creating global branch, check what side effects using persistence.save have
theoretically it should never create here

dlamoris replied:
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/DefaultBranchPermissionsDelegate.java:
```diff
@@ -191,14 +219,15 @@ public PermissionUpdateResponse updateUserPermissions(PermissionUpdateRequest re
                 for (PermissionUpdateRequest.Permission p: req.getPermissions()) {
                     Pair<Group, Role> pair = getGroupAndRole(p);
-                    if (pair.getFirst() == null || pair.getSecond() == null) {
+                    if (pair == null) {
```
check return

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/DefaultBranchPermissionsDelegate.java:
```diff
@@ -225,7 +254,7 @@ public PermissionUpdateResponse updateGroupPermissions(PermissionUpdateRequest r
                 for (PermissionUpdateRequest.Permission p: req.getPermissions()) {
                     Pair<Group, Role> pair = getGroupAndRole(p);
-                    if (pair.getFirst() == null || pair.getSecond() == null) {
+                    if (pair == null) {
```
check return

dlamoris commented on permissions/src/main/java/org/openmbee/mms/permissions/delegation/DefaultPermissionsDelegateFactory.java:
```diff
@@ -1,36 +0,0 @@
-package org.openmbee.mms.permissions.delegation;
```
moved into federated module

Commit:
use returned instance from spring data repository save

dlamoris commented on ldap/src/main/java/org/openmbee/mms/ldap/LdapSecurityConfig.java:
```diff
@@ -196,28 +200,29 @@ public Collection<? extends GrantedAuthority> getGrantedAuthorities(
+                userPersistence.save(user);
+                //Add groups to user
+
+                Set<GroupJson> addGroups = new HashSet<>();
```
will fix in a separate issue/pr

Commit:
fix branch not found behavior when init branch perms

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/DefaultOrgPermissionsDelegate.java:
```diff
@@ -71,6 +74,17 @@ public boolean hasPermission(String user, Set<String> groups, String privilege)
     }
 
+    @Override
+    public boolean hasGroupPermissions(String group, String privilege) {
```
this is actually not used anywhere, just here to fulfill an interface

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/AbstractDefaultPermissionsDelegate.java:
```diff
@@ -60,7 +60,7 @@ protected Pair<Group, Role> getGroupAndRole(PermissionUpdateRequest.Permission p
         Optional<Role> role = getRoleRepo().findByName(p.getRole());
         if (!role.isPresent()) {
-            return Pair.of(group.orElse(null), null);
+            return null;
```
line 67 should keep the returned instance

dlamoris replied:
fixed in #260

Commit:
remove unused things in FederatedPermissionUpdatesResponseBuilder
fix argument in DefaultPermissionService.setProjectInherit

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/FederatedPermissionUpdatesResponseBuilder.java:
```diff
@@ -0,0 +1,76 @@
+    private FederatedPermissionsUpdateResponseBuilder groupsBuilder = new FederatedPermissionsUpdateResponseBuilder();
+
+    @Override
+    public FederatedPermissionUpdatesResponseBuilder setInherit(Boolean inherit) {
```
setInherit, setPublic, insert functions are never called on this class, only on parent class, can be removed

dlamoris replied:
fixed in #260

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/FederatedPermissionsUpdateResponseBuilder.java:
```diff
@@ -0,0 +1,99 @@
+package org.openmbee.mms.federatedpersistence.permissions;
```
these were functions from PermissionsUpdateResponseBuilder that uses the DB classes

dlamoris commented on core/src/main/java/org/openmbee/mms/core/builders/PermissionUpdateResponseBuilder.java:
```diff
@@ -50,93 +44,6 @@ public PermissionUpdateResponseBuilder insert(PermissionUpdateResponse updateUse
         return this;
     }
 
-    public void insertPermissionUpdates_OrgUserPerm(PermissionUpdateResponse.Action action, Collection<OrgUserPerm> perms) {
```
these are moved to federated module

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/DefaultProjectPermissionsDelegate.java:
```diff
@@ -88,6 +85,17 @@ public boolean hasPermission(String user, Set<String> groups, String privilege)
     }
 
+    @Override
+    public boolean hasGroupPermissions(String group, String privilege) {
```
never called, just implements interface

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/permissions/DefaultProjectPermissionsDelegate.java:
```diff
@@ -120,15 +128,28 @@ public boolean setInherit(boolean isInherit) {
     }
 
+    @Override
+    public PermissionResponse getInherit() {
```
never called

Commit:
remove unneeded element delete processing

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/domain/NodeChangeDomain.java:
```diff
@@ -0,0 +1,125 @@
+    }
+
+    public void processElementDeleted(NodeChangeInfo info, ElementJson element) {
+        //TODO Probably don't need this (see method below)
```
removed in #260

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java:
```diff
@@ -77,129 +57,107 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
-        if (b.isDeleted()) {
-            throw new DeletedException(branchesResponse);
+        response.getRefs().add(refJsonOptional.get());
+        if (refJsonOptional.get().isDeleted()) {
```
yes

dlamoris commented on crud/src/main/java/org/openmbee/mms/crud/services/DefaultBranchService.java:
```diff
@@ -77,129 +57,107 @@ public void setEventPublisher(Collection<EventService> eventPublisher) {
-            eventPublisher.forEach((pub) -> pub.publish(
-                EventObject.create(projectId, branch.getId(), "branch_created", branch)));
+            RefJson committedBranch = branchPersistence.save(branch);
+            nodePersistence.branchElements(parentRefOption.get(), parentCommit.get(), committedBranch);
```
fixed in #260

dlamoris commented on localuser/src/main/java/org/openmbee/mms/localuser/controllers/LocalUserController.java:
```diff
@@ -44,10 +48,10 @@ public UserCreateRequest createUser(@RequestBody UserCreateRequest req) {
 
     @GetMapping(value = "/users")
-    @PreAuthorize("isAuthenticated()")
+    @PreAuthorize(AuthorizationConstants.IS_MMSADMIN)
```
user read should not be restricted to admin, reverted in #260

Commit:
users list is available to authenticated users

Commit:
tweak commit update in case twc module is used

Commit:
commit fail cleanup

Commit:
fix commit delete

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedCommitPersistence.java:
```diff
@@ -0,0 +1,183 @@
+        } catch (Exception e) {
+            throw new InternalErrorException("Could not delete commit");
+        }
+        throw new NotFoundException("Could not delete commit");
```
fixed in #260

Commit:
in FederatedNodeChangeDomain.primeNodeChangeInfo, add to any existing info instead of reinit data

Commit:
only further process transacted elements but keep other existing info from passed in nodeChangeInfo

dlamoris commented on federatedpersistence/src/main/java/org/openmbee/mms/federatedpersistence/dao/FederatedNodePersistence.java:
```diff
@@ -0,0 +1,260 @@
+            throw new InternalErrorException("Invalid NodeChangeInfo type presented to NodeFederatedDAO");
+        }
+        ContextHolder.setContext(info.getCommitJson().getProjectId(), info.getCommitJson().getRefId());
+        primeNodeChangeInfo(info, jsons);
```
changed in #260 to not reinit but add to existing data in info

dlamoris committed:
Merge pull request #260 from Open-MBEE/fix/persistence_refactor_fixes

Fix/persistence refactor fixes
